src/Provider.js

```diff
@@ -1,40 +1,35 @@
-import { Children, Component, createContext, createElement } from "react"
+import React from "react"
 import { shallowEqual } from "./utils/utils"
 
-export const MobXProviderContext = createContext({})
+export const MobXProviderContext = React.createContext({})
 
-export class Provider extends Component {
-    static contextType = MobXProviderContext
-    static displayName = "MobXProvider"
-
-    constructor(props, context) {
-        super(props, context)
-        this.state = {
-            ...context,
+export function Provider(props) {
+    const parentValue = React.useContext(MobXProviderContext)
+    const valueRef = React.useRef(null)
+    if (valueRef.current === null) {
```

Comment:
Why to actually use that condition here? If you move that init value inside the `useRef` call, the result will be the same imo.

Reply:
I do not want to create this object on each render. We need to create it once. `React.useRef` does not accept a callback as `React.useState`, so I have to use the condition.

Reply:
Well, in that case, the usual trick is `const [value] = React.useState(() => ({ ... }))`.

Reply:
That was the first version of this code. But I intentionally replaced state with ref, because I am sure that the state is not appropriate here, because we will not change them. It is some static data, not mutable state.

Reply:
Haha, look at `mobx-react-lite` hooks, all based on `useState` even though values don't change. Let's do the optimal now, not because it doesn't feel right.

Reply:
Why do you think that `useState` is the optimal way?

Reply:
And in what way is that optimal? What are the criteria?

Reply:
Ok, optimal was a wrong word ... less confusing will be better. It doesn't make immediate sense why you set ref to null, then check for null and set its value. With `useState` you are just creating a one time state and since there is no setter, it's apparent it's a constant value.

Reply:
Or if you have a problem with useState, then really just set value to `useRef`. Note that React is creating so many objects on each render and throwing away. Creating an extra object in Provider which in most cases doesn't even re-render is less than micro-optimization in cost of worse readability.

Reply:
> Ok, optimal was a wrong word ... less confusing is better. It doesn't make immediate sense why you set ref to null, then check for null and set its value.

This is not something completely new. You can see this idiom in the official Hooks documentation, for example:
https://reactjs.org/docs/hooks-faq.html#how-to-create-expensive-objects-lazily
Other example:
https://github.com/facebook/react/issues/14490#issuecomment-454973512

> Or if you have a problem with useState, then really just set value to useRef. Note that React is creating so many objects on each render and throwing away. Creating an extra object in Provider which in most cases doesn't even re-render is less than micro-optimization in cost of worse readability.

It not about performance only, but about correctness too. We throw an error in the development mode if the stores set changed, but without saving the original value of stores we will rerender with the new set in the production mode and this will be more confusing than rerendering with the original set as we do now.

Reply:
> Or if you have a problem with useState, then really just set value to useRef. Note that React is creating so many objects on each render and throwing away. Creating an extra object in Provider which in most cases doesn't even re-render is less than micro-optimization in cost of worse readability.
 
OK. I will do it. Thanks.

Reply:
> This is not something completely new. You can see this idiom in the official Hooks documentation

Notice the title "How to create **expensive** objects lazily?". We are far from that in here so patterns like that irrelevant, only make code harder to comprehend.

Reply:
> Or if you have a problem with useState, then really just set value to useRef. Note that React is creating so many objects on each render and throwing away. Creating an extra object in Provider which in most cases doesn't even re-render is less than micro-optimization in cost of worse readability.

It is done.